distsql: support local aggregation for AVG

RaduBerinde opened the pull request:
One of the powerful aspects of distsql is that it can perform partial
aggregation locally; for example, if we want to get a sum of values, each node
calculates a local partial sum and then a final node adds these up.

The infrastructure for this was simplistic, in that it only allowed a single
intermediary value. `AVG` is a common aggregation for which that is not
sufficient. The lack of support means that any query using `AVG` won't perform
any local partial aggregation.

We improve the infrastructure and the planning to allow multiple aggregations
along with a final rendering expression. For AVG we compute a sum and a count
and divide them at the end.

The new infrastructure can be used for `VARIANCE` and `STDDEV` as well (the
formulas are more complicated); this is left for future work.

Updates #14351.

@arjunravinarayan I will wait for #14300 to go in and merge with that change.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14637)
<!-- Reviewable:end -->

RaduBerinde committed:
distsql: support local aggregation for AVG

One of the powerful aspects of distsql is that it can perform partial
aggregation locally; for example, if we want to get a sum of values, each node
calculates a local partial sum and then a final node adds these up.

The infrastructure for this was simplistic, in that it only allowed a single
intermediary value. `AVG` is a common aggregation for which that is not
sufficient. The lack of support means that any query using `AVG` won't perform
any local partial aggregation.

We improve the infrastructure and the planning to allow multiple aggregations
along with a final rendering expression. For AVG we compute a sum and a count
and divide them at the end.

The new infrastructure can be used for `VARIANCE` and `STDDEV` as well (the
formulas are more complicated); this is left for future work.

Updates #14351.

arjunravinarayan commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>  Great work! And great testing, the usefulness of the logic test framework is self-evident now.

Do you want to add a TODO about reusing columns? For example, `SELECT url FROM [EXPLAIN (DISTSQL) SELECT AVG(a), SUM(a), COUNT(a) FROM data]` returns https://raduberinde.github.io/decode.html?eJy8lMtrg0AQxu_9K-Q7NbDQ-EiaerL0lEOSksepSNm6gwiJK-sKLcH_vah9RClraIiQQ2Z2v_l94yxzRCoFLfmBcvgvsMHggMEFgweGCUKGTMmI8lyq6kojmIt3-GOGJM0KXaVDhkgqgn-ETvSe4GPL3_a0Ji5IgUGQ5sm-hmQqOXD1EQiuORhWhfatwGb1D2HJIAv9XbZk5wMf41hRzLXs8Da7xW1gj8DwtNott1__66xzknVHZ8N_mUUqlSBFooUMy3_Z2-wWr_PKivNj0G3lvSpaUypIVZ_MurMCxwpcK_CM1u2hB9UDvO6gnKG77QFet1t36G57gNft1hu62x7gcCvnD_ia8kymOXVWT7teyEAipmZr5bJQET0rGdXFm3BV364TgnLdnNpNME-bo8rWqdg2ip2W2O6KHTO5B-0a1Z5Z7F3ie2IUT83k6SXke6N4ZibPLiE_mGc17nkm5kfWZYflzWcAAAD__1vHurI

:dog: I must confess, I have no idea what's going on with the typing code in `aggregator_funcs`. You might want someone else to take a look at that.

---

Reviewed 5 of 7 files at r1, 6 of 7 files at r2.
Review status: 10 of 12 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 827 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2nYrqRXSYkqk4P-AC:-Kh2nYrr_x9ITVe3PHeJ:btu6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsql_physical_planner.go#L827)):*
> ```Go
> 		}
> 
> 		aIdx := 0
> ```

`aIdx` could use a comment, because the way it is incremented by different stages by different amounts is very tricky.

---

*[pkg/sql/distsql_physical_planner.go, line 840 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2o6Hmc7PYtmy4wJqh:-Kh2o6Hmc7PYtmy4wJqi:b-s0n47s) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsql_physical_planner.go#L840)):*
> ```Go
> 					return err
> 				}
> 				intermediateTypes[aIdx] = localResultType
> ```

why the extra `localResultType` variable? If we're returning an error, we don't care that we borked `aIdx`.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2oLscnZlB9A6KgPbm:-Kh2oLscnZlB9A6KgPbn:b5wygu) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsqlplan/aggregator_funcs.go#L39)):*
> ```Go
> 	// The local stage consists of one or more aggregations. All aggregations have
> 	// the same input.
> 	LocalStage []distsqlrun.AggregatorSpec_Func
> ```

Could call it `LocalStages` and `FinalStages`, but I don't insist that you do.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2oUis3jLlsp-FUaB5:-Kh2oUis3jLlsp-FUaB6:b-58s03j) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsqlplan/aggregator_funcs.go#L94)):*
> ```Go
> 	},
> 
> 	distsqlrun.AggregatorSpec_AVG: {
> ```

This could use a high level comment. I can figure this out, but would make it easier going forward, especially if we start doing fancy things with `variance` and things like that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-:-Kh2liCCaJpm3uj5FYft:bmmlz2t)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

The multiple columns thing is part of a bigger problem of how we break things up into pre and post- render expressions. Another symptom is having each tablereader generate constants, e.g. a `2` constant for `2*SUM(x)`. There are a couple of issues open around that, one is #12525.

CC @nvanbenschoten - can you take a quick look over `aggregator_funcs.go`, mainly `FinalRendering`?

---

Review status: 10 of 12 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 827 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2nYrqRXSYkqk4P-AC:-Kh2t4LLaLDjLXpdYP_4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsql_physical_planner.go#L827)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

`aIdx` could use a comment, because the way it is incremented by different stages by different amounts is very tricky.
</blockquote></details>

Done.

---

*[pkg/sql/distsql_physical_planner.go, line 840 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2o6Hmc7PYtmy4wJqh:-Kh2r4yi00-9Sap05Vsr:b-imsclf) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsql_physical_planner.go#L840)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

why the extra `localResultType` variable? If we're returning an error, we don't care that we borked `aIdx`.
</blockquote></details>

Because `:=`. It's either this or `var err error`.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2oLscnZlB9A6KgPbm:-Kh2rDXkFPEgl6lp3uEy:bnmzue5) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsqlplan/aggregator_funcs.go#L39)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Could call it `LocalStages` and `FinalStages`, but I don't insist that you do.
</blockquote></details>

The multiple aggregations together form one stage. If we call them separate stages, it can imply that they feed into each other which is not the case.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 94 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2oUis3jLlsp-FUaB5:-Kh2rZovLD4_678f0cs3:b-l4ult) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsqlplan/aggregator_funcs.go#L94)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

This could use a high level comment. I can figure this out, but would make it easier going forward, especially if we start doing fancy things with `variance` and things like that.
</blockquote></details>

Done, good idea.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-:-Kh2q2xCqOahNbRc_qDX:b-qqmvmr)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Reviewed 1 of 8 files at r3, 1 of 7 files at r4.
Review status: 5 of 12 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 840 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2o6Hmc7PYtmy4wJqh:-Kh3R7U4NGxy7gGEX2G0:b-jnts55) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsql_physical_planner.go#L840)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Because `:=`. It's either this or `var err error`.
</blockquote></details>

fair enough.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2oLscnZlB9A6KgPbm:-Kh3R9vSXJzbKH5SpBln:b-gxztsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsqlplan/aggregator_funcs.go#L39)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The multiple aggregations together form one stage. If we call them separate stages, it can imply that they feed into each other which is not the case.
</blockquote></details>

Fair enough.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 💯

---

Review status: 5 of 12 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 806 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh38CeVvv6hltHkvvpm:-Kh38CeVvv6hltHkvvpn:b8ohrdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/25db5be891ce19f94e15396c72c5d76cacf3cb3e/pkg/sql/distsql_physical_planner.go#L806)):*
> ```Go
> 	} else {
> 		// Some aggregations might need multiple local and final aggregations (with
> 		// a final render expression to get the final result); count of
> ```

I found this comment confusing :)
Can you move "count of aggregations" to the beginning of the sentence? Or make the tone imperative: "Count the number of local aggregators across all the aggregations expressions and keep track of whether any of them needs a final rendering expr".

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2oLscnZlB9A6KgPbm:-Kh39wvsJ2GpxmpPF1-z:buydpfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsqlplan/aggregator_funcs.go#L39)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The multiple aggregations together form one stage. If we call them separate stages, it can imply that they feed into each other which is not the case.
</blockquote></details>

On a related note, I think it might be helpful if we explained more what `DistAggregationInfo`is. Explaining that it's not a part of the logical or the physical plan; instead it's a blueprint of how to create one (or two) stages of the physical plan. (in fact, maybe we should call it blueprint, I like this word now that I've come up with it :) )
This would also help clarify the comment below on FinalRendering about how it could be a "canonical non-typed expression" - that made me question if I understand anything for a second.

Also anything you can do to make it more clear what these arrays are would be helpful. Perhaps just saying more stuff about how an AVG.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh39GB3JBQqyzMkn95J:-Kh39GB4PP1Zu15ZMdWM:bu6nld7) ([raw file](https://github.com/cockroachdb/cockroach/blob/25db5be891ce19f94e15396c72c5d76cacf3cb3e/pkg/sql/distsqlplan/aggregator_funcs.go#L35)):*
> ```Go
> // A simple example is COUNT: the local stage counts (COUNT), the final stage
> // adds the counts (SUM_INT). Other simple examples are SUM, MAX, MIN, boolean
> // operations. A more complex example is AVG.
> ```

well don't tease a brother - say more words on AVG :) Or link to where there's more explanation later in the file.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 49 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh3LCq267yKpHvqkRlF:-Kh3LCq267yKpHvqkRlG:b-n1chfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/25db5be891ce19f94e15396c72c5d76cacf3cb3e/pkg/sql/distsqlplan/aggregator_funcs.go#L49)):*
> ```Go
> 	// if there is more than one aggregation in the stages.
> 	//
> 	// Instead of defining a canonical non-typed expression and then tweaking it
> ```

perhaps expand a little on the non-typed expressions thing: (I'm jumping ahead without reading any code so maybe this suggestion is not correct)
```
// Conceptually this is an expression, to be run as the PostProcessing step of the (single) final stage processor. However, since the DistAggregationInfo is a blueprint (template?) to be used for aggregating inputs of different types, we can't actually use an Expr. We could define a canonical ..., but instead we use a function that acts as an expression generator.
```
And perhaps it should be renamed to `FinalRenderingGenerator`.

But actually - wouldn't having a "canonical non-types expr" just mean we'd have an expression with untyped placeholders in it? I think we already use expressions with placeholders (albeit perhaps typed ones) in other places in distsql (not to mention in prepared statements). Why aren't we doing that here, exactly? If it's not difficult, I think it'd be better because it'd make the definition of a `DistAggregationInfo` declarative and looking at the text of the expression in the source code would immediately tell you what the `FinalStage` entries represent.
At least conceptually it seems to me like it shouldn't be hard... Or at least it's the kind of thing @knz would know how to implement :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-:-Kh3SDuRgVw9Jcs9phO8:b-vprk1r)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented on pkg/sql/testdata/logic_test/aggregate:
```diff
@@ -12,7 +12,7 @@ CREATE TABLE kv (
 query IIIIRRRRBB
 SELECT MIN(1), MAX(1), COUNT(1), SUM_INT(1), AVG(1), SUM(1), STDDEV(1), VARIANCE(1), BOOL_AND(true), BOOL_AND(false) FROM kv
```
It's strange that `SUM_INT` now behaves differently than `SUM`. Also, a lot of these are inconsistent with classic SQL. Are we aware of this, and is it intentional?

```
postgres=# SELECT SUM(1);
 sum
-----
   1
(1 row)
```

nvanbenschoten commented on pkg/sql/distsqlplan/aggregator_funcs.go:
```diff
@@ -16,54 +16,142 @@
+	FinalStage []distsqlrun.AggregatorSpec_Func
+
+	// An optional rendering expression used to obtain the final result; required
+	// if there is more than one aggregation in the stages.
```
nit: `s/in the stages/in each of the stages`

nvanbenschoten commented on pkg/sql/distsqlplan/aggregator_funcs.go:
```diff
@@ -16,54 +16,142 @@
+			}
+
+			if sum.ResolvedType().Equivalent(parser.TypeFloat) {
+				// There is no "FLOAT / INT" operator; cast the denominator to float.
```
Might want to mention that there is a "DECIMAL/INT" operator, so we don't need the same handling for that case.

nvb reviewed (commented):
`aggregator_funcs.go` looks fine. Very cool stuff here!

RaduBerinde commented:
Review status: 5 of 12 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql_physical_planner.go, line 806 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh38CeVvv6hltHkvvpm:-Kh3ypo-U-21Hl5nq02M:b-w098yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/25db5be891ce19f94e15396c72c5d76cacf3cb3e/pkg/sql/distsql_physical_planner.go#L806)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I found this comment confusing :)
Can you move "count of aggregations" to the beginning of the sentence? Or make the tone imperative: "Count the number of local aggregators across all the aggregations expressions and keep track of whether any of them needs a final rendering expr".
</blockquote></details>

Done. Something was definitely wrong in that comment

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh2oLscnZlB9A6KgPbm:-Kh3X0290J7QzK9zoPFT:bframe9) ([raw file](https://github.com/cockroachdb/cockroach/blob/231c1536b2dbbe346707c82f79cafb0c4c9f83c1/pkg/sql/distsqlplan/aggregator_funcs.go#L39)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

On a related note, I think it might be helpful if we explained more what `DistAggregationInfo`is. Explaining that it's not a part of the logical or the physical plan; instead it's a blueprint of how to create one (or two) stages of the physical plan. (in fact, maybe we should call it blueprint, I like this word now that I've come up with it :) )
This would also help clarify the comment below on FinalRendering about how it could be a "canonical non-typed expression" - that made me question if I understand anything for a second.

Also anything you can do to make it more clear what these arrays are would be helpful. Perhaps just saying more stuff about how an AVG.
</blockquote></details>

I like "blueprint" :) Added comments. Will rename to Blueprint in a separate commit when the PR quiesces.

Do you think that instead of having two arrays of the same length, it would be more clear to have a single array for the "inner" aggregations, `[]struct { Local, Final distsqlrun.AggregatorSpec_func }` ?

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 35 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh39GB3JBQqyzMkn95J:-Kh3zAjlm6fQt4KdVHrz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25db5be891ce19f94e15396c72c5d76cacf3cb3e/pkg/sql/distsqlplan/aggregator_funcs.go#L35)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well don't tease a brother - say more words on AVG :) Or link to where there's more explanation later in the file.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 49 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kh3LCq267yKpHvqkRlF:-Kh3VnaFoLSvF7_43BPw:b-xbc51r) ([raw file](https://github.com/cockroachdb/cockroach/blob/25db5be891ce19f94e15396c72c5d76cacf3cb3e/pkg/sql/distsqlplan/aggregator_funcs.go#L49)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

perhaps expand a little on the non-typed expressions thing: (I'm jumping ahead without reading any code so maybe this suggestion is not correct)
```
// Conceptually this is an expression, to be run as the PostProcessing step of the (single) final stage processor. However, since the DistAggregationInfo is a blueprint (template?) to be used for aggregating inputs of different types, we can't actually use an Expr. We could define a canonical ..., but instead we use a function that acts as an expression generator.
```
And perhaps it should be renamed to `FinalRenderingGenerator`.

But actually - wouldn't having a "canonical non-types expr" just mean we'd have an expression with untyped placeholders in it? I think we already use expressions with placeholders (albeit perhaps typed ones) in other places in distsql (not to mention in prepared statements). Why aren't we doing that here, exactly? If it's not difficult, I think it'd be better because it'd make the definition of a `DistAggregationInfo` declarative and looking at the text of the expression in the source code would immediately tell you what the `FinalStage` entries represent.
At least conceptually it seems to me like it shouldn't be hard... Or at least it's the kind of thing @knz would know how to implement :)
</blockquote></details>

The expression needs to refer to its inputs via `IndexedVar`s (not placeholders), because that's what the machinery to serialize it for distsql expects. The indices of the `IndexedVar`s we want to serialize are not set in stone (they depend on where the `AVG` or whatever is in the sequence of aggregations), so we would need some kind of visitor to go through and remap those. Even with placeholders I think we would need that remapping.

In addition, because of the `FLOAT` / `INT` complication, we actually don't have a single type of expression.

So yeah, it would be way more difficult than this. Improved the comment.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 121 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kgz9dztT5D5UIvACpfs-r3-121:-Kh415E6EY2pab-Kt48C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/25db5be891ce19f94e15396c72c5d76cacf3cb3e/pkg/sql/distsqlplan/aggregator_funcs.go#L121)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Might want to mention that there is a "DECIMAL/INT" operator, so we don't need the same handling for that case.
</blockquote></details>

Done.

---

*[pkg/sql/testdata/logic_test/aggregate, line 13 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kgz9dzuhnlqnZEqCv0w-r4-13:-Kh4164zorXQDS9NzaVt:b-hquceo) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c763bf4dc2495870885275b900298c92ccd57e/pkg/sql/testdata/logic_test/aggregate#L13)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's strange that `SUM_INT` now behaves differently than `SUM`. Also, a lot of these are inconsistent with classic SQL. Are we aware of this, and is it intentional?

```
postgres=# SELECT SUM(1);
 sum
-----
   1
(1 row)
```
</blockquote></details>

SUM_INT was defined only for this purpose so I figured it's fine for it to always be an int as the name suggests, but if you feel the consistency is more important I can revert this part, it should work both ways.

As for the other issue, indeed that seems wrong. I posted this case in #12525.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 3 of 12 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/testdata/logic_test/aggregate, line 13 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kgz9dzuhnlqnZEqCv0w-r4-13:-Kh47qMYyM_Jndt2ucNC:b-y2rp6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c763bf4dc2495870885275b900298c92ccd57e/pkg/sql/testdata/logic_test/aggregate#L13)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

SUM_INT was defined only for this purpose so I figured it's fine for it to always be an int as the name suggests, but if you feel the consistency is more important I can revert this part, it should work both ways.

As for the other issue, indeed that seems wrong. I posted this case in #12525.
</blockquote></details>

SUM_INT is Cockroach-specific, so consistency isn't quite as important. Still, I think it's important for all of our aggregates to behave the same when run over 0 rows, otherwise we can't document a concrete rule. Can't we just catch NULL and convert it to 0 where necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 3 of 12 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/testdata/logic_test/aggregate, line 13 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kgz9dzuhnlqnZEqCv0w-r4-13:-Kh49bWWPawNFdIyBQJg:btttd23) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c763bf4dc2495870885275b900298c92ccd57e/pkg/sql/testdata/logic_test/aggregate#L13)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

SUM_INT is Cockroach-specific, so consistency isn't quite as important. Still, I think it's important for all of our aggregates to behave the same when run over 0 rows, otherwise we can't document a concrete rule. Can't we just catch NULL and convert it to 0 where necessary?
</blockquote></details>

The problem is when `COUNT`ing 0 rows. The result of the count in a distributed aggregation is the result of `SUM_INT`, and we want the result to be 0 not NULL. Maybe we want to rename it to something like `COUNT_SUM()`? (or `CRDB_INTERNAL.COUNT_SUM`)

This was working before because each local stage was emitting a result even if there were no rows, so SUM_INT was adding up 0s; but this had to be corrected because of the IDENT stuff.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: ignore NULLs in ident aggregator

Some aggregations pass through constant values using IDENT functions. If one of
the local stages produces no results, the resulting row contains a NULL; this
NULL shouldn't mess up the result of the final IDENT aggregation. Change the
aggregator to ignore NULLs (unless there are only NULLs).

RaduBerinde commented:
*[pkg/sql/testdata/logic_test/aggregate, line 13 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-Kgz9dzuhnlqnZEqCv0w-r4-13:-Kh4atxNbO701pdtLH4w:b-lh3moy) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c763bf4dc2495870885275b900298c92ccd57e/pkg/sql/testdata/logic_test/aggregate#L13)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The problem is when `COUNT`ing 0 rows. The result of the count in a distributed aggregation is the result of `SUM_INT`, and we want the result to be 0 not NULL. Maybe we want to rename it to something like `COUNT_SUM()`? (or `CRDB_INTERNAL.COUNT_SUM`)

This was working before because each local stage was emitting a result even if there were no rows, so SUM_INT was adding up 0s; but this had to be corrected because of the IDENT stuff.
</blockquote></details>

I have reworked the second commit; instead of not emitting these local stage rows, we now ignore NULLs in the IDENT aggregator. PTAL

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Second commit LGTM. I think it's better than the previous alternative. Let's link to the previous hack in 70b6815 in the TODO, because both are working around the same issue. Also, do we have an open issue to get rid of the ident aggregator?

RaduBerinde commented:
Thanks. Added to the TODO. #12525 now tracks this.

---

Review status: 2 of 10 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14637#-:-KhOZvL3zp2Ohfet34Bl:b4gva0y)*
<!-- Sent from Reviewable.io -->
